### PR TITLE
Md rails issue 5, 6 and 7

### PR DESCRIPTION
**Pull request by mertdusunceli**

### CODE REVIEW

**Comment by mertdusunceli** [comment]
[Issue 5: Image Tags](https://trello.com/c/HaGp7qe7/311-issue-5-image-tags)

**Comment by mertdusunceli** [comment]
[Issue 6: Images by Tag](https://trello.com/c/NRWHw4l8/310-issue-6-images-by-tag)

**Comment by mertdusunceli** [comment]
[Issue 6: Images by Tag](https://trello.com/c/NRWHw4l8/310-issue-6-images-by-tag)

**Comment by mertdusunceli** [comment]
[Issue 7: Delete Image](https://trello.com/c/zGgEAz3U/314-issue-7-delete-image)

File: test/controllers/images_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+require 'test_helper'
+
+class ImagesControllerTest < ActionDispatch::IntegrationTest
+  test 'should get index page' do
+    get '/'
+    assert_response :success
```
Comment (eric199wen) [inline test/controllers/images_controller_test.rb]: We would like to use `:ok`. for `assert_response` since it is more specific.

Comment (eric199wen) [reply test/controllers/images_controller_test.rb]: We also like to assert on response body to make sure we hit the right template.

File: test/controllers/images_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      post '/images', params: { image: { image_url: url } }
+    end
+    assert_response :redirect
+    follow_redirect!
+    assert_response :success
+    assert_includes response.body, url
+  end
+
+  test 'should not create image' do
+    get '/images/new'
```
Comment (eric199wen) [inline test/controllers/images_controller_test.rb]: We don't need this line.

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    assert_no_difference('Image.count') do
+      post '/images', params: { image: { image_url: ' ' } }
+    end
+    assert_response :success
+  end
+
+  test 'should show an image' do
+    url = 'https://upload.wikimedia.org/wikipedia/commons/thumb/9/97/Lockheed_SR-71_Blackbird.jpg/1920px-Lockheed_SR-71_Blackbird.jpg'
+    post '/images', params: { image: { image_url: url } }
```
Comment (eric199wen) [inline test/controllers/images_controller_test.rb]: We could do Image.create! to setup data for testing instead of hitting create action.

**Comment by stephaniemhutson** [comment]
You can fix the rubocop errors that you're getting in ci by adding the taggable migrations to the rubocop.yml file

ex: https://github.com/AppFolioOnboarding/image-sharer-stephaniemhutson/pull/10/commits/212eebd69af73c32c76fb41a0ee4ba45667efd77

There are some errs that exist in your manual changes though, and running rubocop locally will help you identify them - also there might be a plugin for vscode (there is for sublime) which will highlight rubocop errors for you.

File: app/views/images/new.html.erb [hunk 4]
```diff
@@ -1,7 +1,8 @@
 <h3>Add Image</h3>
 
 <%= simple_form_for @image do |f| %>
-  <%= f.input :image_url %>
+  <%= f.input :image_url, :required => false, placeholder: 'Add Image URL', label: false %>
```
Comment (eric199wen) [inline app/views/images/new.html.erb]: We would like to use hash syntax for option param.

File: config/routes.rb [hunk 5]
```diff
@@ -1,4 +1,6 @@
 Rails.application.routes.draw do
   resources :images
+  resources :tags, only: [:index, :show]
```
Comment (eric199wen) [inline config/routes.rb]: We probably won't need this?

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -37,4 +37,36 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
     assert_response :success
     assert_includes response.body, url
   end
+
+  test 'should create image with tags present and jump to tag page' do
+    url = 'https://upload.wikimedia.org/wikipedia/commons/thumb/9/97/Lockheed_SR-71_Blackbird.jpg/1920px-Lockheed_SR-71_Blackbird.jpg'
+    post '/images', params: { image: { image_url: url, tag_list: 'SR-71, plane' } }
```
Comment (eric199wen) [inline test/controllers/images_controller_test.rb]: In unit test, we only need to test the action to keep the scope small. If we want to do integration test, we could have selenium test.

File: test/controllers/images_controller_test.rb [hunk 7]
```diff
@@ -37,4 +37,36 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+    emptyPageTag = '/tagged?tag=empty'
+    get emptyPageTag
+    expected = '<h6>Currently at "empty" tag</h6>'
+    assert_includes response.body, expected
+  end
+
+  test 'should raise error from empty tag page' do
+    blankTag = '/tagged?tag='
+    exception = assert_raises(Exception) {get blankTag}
+    assert_includes exception.message, "nil"
```
Comment (stephaniemhutson) [inline test/controllers/images_controller_test.rb]: "nil" is not particularly descriptive, are you able to do an assert_equal on the exception message?

File: app/views/images/new.html.erb (hunk 4 above)
Comment (mertdusunceli) [reply app/views/images/new.html.erb]: Done!

File: config/routes.rb (hunk 5 above)
Comment (mertdusunceli) [reply config/routes.rb]: Done!

File: test/controllers/images_controller_test.rb (hunk 1 above)
Comment (mertdusunceli) [reply test/controllers/images_controller_test.rb]: Done!

File: test/controllers/images_controller_test.rb (hunk 3 above)
Comment (mertdusunceli) [reply test/controllers/images_controller_test.rb]: Done!

File: test/controllers/images_controller_test.rb (hunk 6 above)
Comment (mertdusunceli) [reply test/controllers/images_controller_test.rb]: Fixed the tests!

File: test/controllers/images_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    assert_difference('Image.count') do
+      post '/images', params: { image: { image_url: url } }
+    end
+    assert_response :redirect
+    follow_redirect!
+    assert_response :success
+    assert_includes response.body, url
+  end
+
+  test 'should not create image' do
```
[the remark this answers is not in the document]
Comment (mertdusunceli) [reply test/controllers/images_controller_test.rb]: You are right, fixed it!

File: test/controllers/images_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    assert_response :success
+  end
+
+  test 'should get new image page' do
+    get '/images/new'
+    assert_response :success
+  end
+
+  test 'should create image and redirect' do
+    url = 'https://upload.wikimedia.org/wikipedia/commons/thumb/9/97/Lockheed_SR-71_Blackbird.jpg/1920px-Lockheed_SR-71_Blackbird.jpg'
```
[the remark this answers is not in the document]
Comment (mertdusunceli) [reply test/controllers/images_controller_test.rb]: Yes it looks way cleaner now.

File: test/controllers/images_controller_test.rb (hunk 7 above)
Comment (mertdusunceli) [reply test/controllers/images_controller_test.rb]: Fixed it!

**Pull request merged**